### PR TITLE
[prompty] prompty support retry

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
Migrate the existing tool's retry logic to prompty.
reference code: https://github.com/microsoft/promptflow/blob/106401c2e631b95575504e36b6594ee0519e62e7/src/promptflow-tools/promptflow/tools/common.py#L522C1-L523C1

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/core/_errors.py [hunk 1]
```diff
@@ -1,8 +1,79 @@
...
+        msg = "The api key is invalid or revoked. " "You can correct or regenerate the api key of your connection."
+        return f"OpenAI API hits {ex_type}: {msg}"
+    # for models that do not support the `functions` parameter.
+    elif "Unrecognized request argument supplied: functions" in error_message:
+        msg = (
+            "Current model does not support the `functions` parameter. If you are using openai connection, then "
+            "please use gpt-3.5-turbo, gpt-4, gpt-4-32k, gpt-3.5-turbo-0613 or gpt-4-0613. You can refer to "
+            "https://platform.openai.com/docs/guides/gpt/function-calling. If you are using azure openai "
+            "connection, then please first go to your Azure OpenAI resource, deploy model 'gpt-35-turbo' or "
+            "'gpt-4' with version 0613, then go to prompt flow connection page, upgrade connection api version to "
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_errors.py:26]: we might not have a prompt flow connection page in local

File: src/promptflow-core/promptflow/core/_errors.py [hunk 2]
```diff
@@ -1,8 +1,79 @@
...
+        )
+        return f"OpenAI API hits {ex_type}: {msg}"
+    elif ("'response_format' of type" in error_message and "is not supported with this model." in error_message) or (
+        "Additional properties are not allowed" in error_message and "unexpected) - 'response_format'" in error_message
+    ):
+        msg = (
+            'The response_format parameter needs to be a dictionary such as {"type": "text"}. '
+            "The value associated with the type key should be either 'text' or 'json_object' "
+            'If you are using openai connection, you can only set response_format to { "type": "json_object" } '
+            "when calling gpt-3.5-turbo-1106 or gpt-4-1106-preview to enable JSON mode. You can refer to "
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_errors.py:60]: gpt-3.5-turbo-1106 or later?

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 3]
```diff
@@ -448,4 +461,275 @@ def parse_chat(chat_str, images: List = None, valid_roles: List[str] = None):
...
+def refine_extra_fields_not_permitted_error(connection, deployment_name, model):
+    tsg = (
+        "Please kindly avoid using vision model in LLM tool, "
+        "because vision model cannot work with some chat api parameters. "
+        "You can change to use tool 'Azure OpenAI GPT-4 Turbo with Vision' "
+        "or 'OpenAI GPT-4V' for vision model."
+    )
+    try:
+        if isinstance(connection, AzureOpenAIConnection):
+            subscription_id, resource_group, workspace_name = get_workspace_triad()
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:589]: are we alwasy in connection provider = azure here?

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 4]
```diff
@@ -448,4 +461,275 @@ def parse_chat(chat_str, images: List = None, valid_roles: List[str] = None):
...
+        "server disconnected without sending a response",
+    ]
+    for message in retriable_error_messages:
+        if message in str(e).lower() or message in str(e.__cause__).lower():
+            return True
+
+    return False
+
+
+# TODO(2971352): revisit this tries=100 when there is any change to the 10min timeout logic
```
Comment (wangchao1230) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:622]: remove this TODO line

File: src/promptflow-core/promptflow/core/_prompty_utils.py [hunk 5]
```diff
@@ -448,4 +461,275 @@ def parse_chat(chat_str, images: List = None, valid_roles: List[str] = None):
...
+def handle_openai_error(tries: int = 100, unprocessable_entity_error_tries: int = 3):
```
Comment (lalala123123) [inline src/promptflow-core/promptflow/core/_prompty_utils.py:623]: change to 10

Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py:623]: describle max time
add to doc

File: src/promptflow-core/promptflow/core/_errors.py (hunk 1 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_errors.py]: remove it

File: src/promptflow-core/promptflow/core/_errors.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_errors.py:57]: update the description

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: remove it

File: src/promptflow-core/promptflow/core/_prompty_utils.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow-core/promptflow/core/_prompty_utils.py]: remove it, this error message refine is used for LLM tool

**Review by wangchao1230 — APPROVED** [review]

**Review by elliotzh — APPROVED** [review]

**Pull request merged**